### ISSUE TITLE
Different artifact's content after registering an AsyncAPI definition

### ISSUE THREAD

**Issue by rmarting** [Apicurio/apicurio-registry#1787]
I found that registering the same API schema, in my case an AsyncAPI sample, the results are not the same. The results depend of the method you used to register the API, somecases successfully and in others not. 

To give a successful behaviour to the user, each case must work exactly and produce the same results, otherwise it could produce errors for third-party getting the artifact's content.

## Steps to reproduce it

Components, tools and versions used:

- Apicurio Registry Version: 2.0.1 (using the apicurio-registry-mem:2.0.1.Final container image)
- Apicurio Registry Maven Plugin 2.0.1.Final
- Google Chrome 92.0.4515.159 (Official Build) (64-bit)
- Firefox 91.0 (64-bit)

AsyncAPI Sample definition:

```yaml
asyncapi: 2.1.0
info:
  title: Hello world application
  version: '0.1.0'
channels:
  hello:
    publish:
      message:
        payload:
          type: string
          pattern: '^hello .+$'
```

## Case 1: Using Apicurio Registry Maven Plugin

This is the definition to upload the artifact:

```xml
                    <plugin>
                        <groupId>io.apicurio</groupId>
                        <artifactId>apicurio-registry-maven-plugin</artifactId>
                        <version>${apicurio.version}</version>
                        <executions>
                            <execution>
                                <phase>generate-sources</phase>
                                <goals>
                                    <goal>register</goal>
                                </goals>
                                <configuration>
                                    <registryUrl>${apicurio.registry.url}</registryUrl>
                                    <artifacts>
                                        <!-- AsyncAPI Spec -->
                                        <artifact>
                                            <groupId>asyncapi</groupId>
                                            <artifactId>hello-world</artifactId>
                                            <type>ASYNCAPI</type>
                                            <file>
                                                ${project.basedir}/src/main/asyncapi/hello-world.yml
                                            </file>
                                            <ifExists>RETURN_OR_UPDATE</ifExists>
                                            <canonicalize>true</canonicalize>
                                        </artifact>
                                    </artifacts>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
```

**Results:** The artifact is registered successfully. If you review the content in the UI, the content is the same as the original one.

## Case 2: Uploading by UI with Google Chrome

Registering with auto-detection and copying the content in the artifact section.

**Results**: It fails with the following exception in the UI:

```text
io.apicurio.registry.storage.InvalidArtifactTypeException: Failed to discover artifact type from content.
	at io.apicurio.registry.util.ArtifactTypeUtil.discoverType(ArtifactTypeUtil.java:132)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl.determineArtifactType(GroupsResourceImpl.java:639)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl.createArtifact(GroupsResourceImpl.java:519)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl_Subclass.createArtifact$$superaccessor31(GroupsResourceImpl_Subclass.zig:5963)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl_Subclass$$function$$31.apply(GroupsResourceImpl_Subclass$$function$$31.zig:71)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:54)
	at io.smallrye.metrics.interceptors.CountedInterceptor.countedCallable(CountedInterceptor.java:95)
	at io.smallrye.metrics.interceptors.CountedInterceptor.countedMethod(CountedInterceptor.java:70)
	at io.smallrye.metrics.interceptors.CountedInterceptor_Bean.intercept(CountedInterceptor_Bean.zig:366)
	at io.quarkus.arc.impl.InitializedInterceptor.intercept(InitializedInterceptor.java:79)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.smallrye.metrics.interceptors.TimedInterceptor.timedCallable(TimedInterceptor.java:95)
	at io.smallrye.metrics.interceptors.TimedInterceptor.timedMethod(TimedInterceptor.java:70)
	at io.smallrye.metrics.interceptors.TimedInterceptor_Bean.intercept(TimedInterceptor_Bean.zig:366)
	at io.quarkus.arc.impl.InitializedInterceptor.intercept(InitializedInterceptor.java:79)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.smallrye.metrics.interceptors.ConcurrentGaugeInterceptor.concurrentCallable(ConcurrentGaugeInterceptor.java:96)
	at io.smallrye.metrics.interceptors.ConcurrentGaugeInterceptor.countedMethod(ConcurrentGaugeInterceptor.java:69)
	at io.smallrye.metrics.interceptors.ConcurrentGaugeInterceptor_Bean.intercept(ConcurrentGaugeInterceptor_Bean.zig:366)
	at io.quarkus.arc.impl.InitializedInterceptor.intercept(InitializedInterceptor.java:79)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.metrics.RestMetricsInterceptor.intercept(RestMetricsInterceptor.java:82)
	at io.apicurio.registry.metrics.RestMetricsInterceptor_Bean.intercept(RestMetricsInterceptor_Bean.zig:327)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:50)
	at io.apicurio.registry.logging.LoggingInterceptor.logMethodEntry(LoggingInterceptor.java:55)
	at io.apicurio.registry.logging.LoggingInterceptor_Bean.intercept(LoggingInterceptor_Bean.zig:275)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:41)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl_Subclass.createArtifact(GroupsResourceImpl_Subclass.zig:5908)
	at io.apicurio.registry.rest.v2.GroupsResourceImpl_ClientProxy.createArtifact(GroupsResourceImpl_ClientProxy.zig:819)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:408)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at org.jboss.resteasy.plugins.server.servlet.ServletContainerDispatcher.service(ServletContainerDispatcher.java:249)
	at io.quarkus.resteasy.runtime.ResteasyFilter.doFilter(ResteasyFilter.java:35)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apicurio.registry.ui.servlets.ResourceCacheControlFilter.doFilter(ResourceCacheControlFilter.java:83)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.apicurio.registry.rest.RegistryApplicationServletFilter.doFilter(RegistryApplicationServletFilter.java:113)
	at io.apicurio.registry.rest.RegistryApplicationServletFilter_ClientProxy.doFilter(RegistryApplicationServletFilter_ClientProxy.zig:225)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:63)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:67)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:133)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:65)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:247)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:56)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:111)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:108)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$9$1.call(UndertowDeploymentRecorder.java:574)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:227)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleRequest(ServletInitialHandler.java:152)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$1.handleRequest(UndertowDeploymentRecorder.java:117)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:290)
	at io.undertow.server.DefaultExchangeHandler.handle(DefaultExchangeHandler.java:18)
	at io.quarkus.undertow.runtime.UndertowDeploymentRecorder$5$1.run(UndertowDeploymentRecorder.java:400)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

## Case 3: Uploading by UI with Firefox

Registering with auto-detection and copying the content in the artifact section.

**Results:** It is registered successfully, however if you review the content in the UI, the content was reformatted as JSON:

```json
{
  "asyncapi" : "2.1.0",
  "info" : {
    "title" : "Hello world application",
    "version" : "0.1.0"
  },
  "channels" : {
    "hello" : {
      "publish" : {
        "message" : {
          "payload" : {
            "type" : "string",
            "pattern" : "^hello .+$"
          }
        }
      }
    }
  }
}
```

## Summary

To summarize the bug:

* Case 1 :heavy_check_mark: Artifact's content in registry is the same as the original one.
* Case 2: :heavy_multiplication_x: Unable to register the artifact.
* Case 3: :heavy_multiplication_x: Artifact's content in registry is different from the original one.

I did not check other specifications, but I guess that the results could be very similar.

It seems that #1651 issue fixed the Case 2 but I could reproduce again. At the same time the #1659 is related with auto-detection of AsyncAPI spec, so maybe this bug is related to that one.

WDTY?

**Comment by EricWittmann** [comment]
Thank you for the detailed report!  We definitely have some issues with YAML files vs. JSON formatted files.  This is something that needs more extensive testing and you've done a nice job here to get us started.  I don't have anything to add right now - I will need to dig into it and do some of my own testing.

I wonder if perhaps @SafarMirek has any insights into this?

**Comment by rmarting** [comment]
As you said, it seems that not everything is working find with YAML and JSON files, and Apicurio Registry needs to invest and improve it soon. Many different API schemas and two different formats to cover everything ... it sounds not easy. Maybe the solution could be aligned to avoid any kind of reformatting process between YAML or JSON and manage the data as is (or as is provided by the user by UI or Maven), but it is only an idea. You are the experts and I will wait for your analysis and results. 

I am open to help to test or verify any idea about it.

**Comment by SafarMirek** [comment]
Currently, Registry is designed in way that it should always convert yaml into json and UI assumes that, so that's the reason why documentation tab does not work for you in case 1.
There is probably a problem with maven plugin which does not set content type of http request to YAML. 

@rmarting I think if you enabled global validation you would not be able to create that artifact with maven plugin. Please, let me know if I am right or not.

**Comment by rmarting** [comment]
@SafarMirek you are right, if I am enabled global validacion (validation rule - full) then Apicurio Maven Plug-In fails with the following exception:

```log
[ERROR] Exception while registering artifact [io.jromanmartin.kafka.schema.asyncapi] / [Message]
io.apicurio.registry.rest.client.exception.RuleViolationException: Syntax violation for AsyncAPI artifact.
```

With that validation rule enabled I could avoid to upload API definitions not-compliant with Apicuiro, however I got other issue: I have to transform my YAML API definitions to other valid for Apicurio (in our case using JSON). In that case we are basically in the same place, we have a limitation in Apicurio Registry to manage different API schemas formats, if JSON is the only valid. 

Could Apicurio Maven Plug-In transform my YALM definition to JSON at registering time, as UI does? I will be a great improvement to help developers to work with Apicurio Registry.

Ideas?

**Comment by SafarMirek** [comment]
@rmarting Thanks for testing it. 
Yeah, it's a bug in maven plugin and it needs to be fixed. Maven plugin should work same way as the UI. 

I will look at it.

**Comment by SafarMirek** [comment]
@rmarting Hi, can you try your tests with my pull request?
Thanks

**Comment by rmarting** [comment]
@SafarMirek I built and installed locally your branch into my local Maven Repository as 2.1.0-SNAPSHOT to test and verify the Maven plugin. I also deployed the local instance of Apicuiro using your branch and the command `quarkus:dev`.

My test results are:

* Disabled Global Validation Rules: API-YAML definitions are registered but the content in Apicurio Registry are still YAML.
* Enabled Global Validation Rules: API-YAML definitions fails with the next exception:

```log
[ERROR] Exception while registering artifact [io.jromanmartin.kafka.schema.asyncapi] / [Message]
io.apicurio.registry.rest.client.exception.RuleViolationException: Syntax violation for AsyncAPI artifact.
```

HTH

**Comment by SafarMirek** [comment]
@rmarting I have tested it with the current version of my PR and it looks it works,  but you are free to test it out if you want.
